File: see/beetle12.pl
% beetle example 12

:- use_module(library(iso_ext)).
:- use_module(library(lists)).
:- use_module(library(terms)).

:- dynamic(brake/0).
:- dynamic('https://eyereasoner.github.io/see#sequent'/2).
:- dynamic('urn:example:is'/2).

% beetle is a car
'urn:example:Car'('urn:example:beetle').

% all cars are green or blue
'https://eyereasoner.github.io/see#sequent'(
    'urn:example:Car'(A),
    [
        'urn:example:is'(A,'urn:example:green'),
        'urn:example:is'(A,'urn:example:blue')
    ]
).

% green things are nice or pretty
'https://eyereasoner.github.io/see#sequent'(
    'urn:example:is'(A,'urn:example:green'),
    [
        'urn:example:is'(A,'urn:example:nice'),
        'urn:example:is'(A,'urn:example:pretty')
    ]
).

% pretty things are pretty1 or pretty2
'https://eyereasoner.github.io/see#sequent'(
    'urn:example:is'(A,'urn:example:pretty'),
    [
        'urn:example:is'(A,'urn:example:pretty1'),
        'urn:example:is'(A,'urn:example:pretty2')
    ]
).

% nice things are nice1 or nice2
'https://eyereasoner.github.io/see#sequent'(
    'urn:example:is'(A,'urn:example:nice'),
    [
        'urn:example:is'(A,'urn:example:nice1'),
        'urn:example:is'(A,'urn:example:nice2')
    ]
).

% pretty1 things are pretty11 or pretty12
'https://eyereasoner.github.io/see#sequent'(
    'urn:example:is'(A,'urn:example:pretty1'),
    [
        'urn:example:is'(A,'urn:example:pretty11'),
        'urn:example:is'(A,'urn:example:pretty12')
    ]
).

% pretty2 things are pretty21 or pretty22
'https://eyereasoner.github.io/see#sequent'(
    'urn:example:is'(A,'urn:example:pretty2'),
    [
        'urn:example:is'(A,'urn:example:pretty21'),
        'urn:example:is'(A,'urn:example:pretty22')
    ]
).

% nice1 things are nice11 or nice12
'https://eyereasoner.github.io/see#sequent'(
    'urn:example:is'(A,'urn:example:nice1'),
    [
        'urn:example:is'(A,'urn:example:nice11'),
        'urn:example:is'(A,'urn:example:nice12')
    ]
).

% nice2 things are nice21 or nice22
'https://eyereasoner.github.io/see#sequent'(
    'urn:example:is'(A,'urn:example:nice2'),
    [
        'urn:example:is'(A,'urn:example:nice21'),
        'urn:example:is'(A,'urn:example:nice22')
    ]
).

% pretty or nice or blue things are beautiful
'https://eyereasoner.github.io/see#sequent'(
    'urn:example:is'(A,'urn:example:pretty11'),
    [
        'urn:example:is'(A,'urn:example:beautiful')
    ]
).

'https://eyereasoner.github.io/see#sequent'(
    'urn:example:is'(A,'urn:example:pretty12'),
    [
        'urn:example:is'(A,'urn:example:beautiful')
    ]
).

'https://eyereasoner.github.io/see#sequent'(
    'urn:example:is'(A,'urn:example:pretty21'),
    [
        'urn:example:is'(A,'urn:example:beautiful')
    ]
).

'https://eyereasoner.github.io/see#sequent'(
    'urn:example:is'(A,'urn:example:pretty22'),
    [
        'urn:example:is'(A,'urn:example:beautiful')
    ]
).

'https://eyereasoner.github.io/see#sequent'(
    'urn:example:is'(A,'urn:example:nice11'),
    [
        'urn:example:is'(A,'urn:example:beautiful')
    ]
).

'https://eyereasoner.github.io/see#sequent'(
    'urn:example:is'(A,'urn:example:nice12'),
    [
        'urn:example:is'(A,'urn:example:beautiful')
    ]
).

'https://eyereasoner.github.io/see#sequent'(
    'urn:example:is'(A,'urn:example:nice21'),
    [
        'urn:example:is'(A,'urn:example:beautiful')
    ]
).

'https://eyereasoner.github.io/see#sequent'(
    'urn:example:is'(A,'urn:example:nice22'),
    [
        'urn:example:is'(A,'urn:example:beautiful')
    ]
).

'https://eyereasoner.github.io/see#sequent'(
    'urn:example:is'(A,'urn:example:blue'),
    [
        'urn:example:is'(A,'urn:example:beautiful')
    ]
).

'https://eyereasoner.github.io/see#sequent'(
    'urn:example:is'('urn:example:beetle','urn:example:beautiful'),
    [
        'urn:example:is'('urn:example:test',true)
    ]
).

% resolution
'https://eyereasoner.github.io/see#sequent'(
    (
        'https://eyereasoner.github.io/see#sequent'(A,B),
        select(C,B,D),
        'https://eyereasoner.github.io/see#sequent'(C,E),
        length(E,F),
        F =< 1,
        append(E,D,G)
    ),
    [
        'https://eyereasoner.github.io/see#sequent'(A,G)
    ]
).

% factoring
'https://eyereasoner.github.io/see#sequent'(
    (
        'https://eyereasoner.github.io/see#sequent'(A,B),
        list_to_set(B,C)
    ),
    [
        'https://eyereasoner.github.io/see#sequent'(A,C)
    ]
).

% sequent
sequent :-
    (   'https://eyereasoner.github.io/see#sequent'(A,Bs),
        A,
        (   Bs = []
        ->  write(fuse('https://eyereasoner.github.io/see#sequent'(A,Bs),A)),
            nl,
            halt(2)
        ;   true
        ),
        Bs = [B],
        \+B,
        step(B),
        retract(brake),
        false
    ;   brake,
        !
    ;   assertz(brake),
        sequent
    ).

step((A,B)) :-
    !,
    step(A),
    step(B).
step(A) :-
    (   \+A
    ->  assertz(A)
    ;   true
    ).

% query
query('urn:example:is'(_A,_B)).

test :-
    sequent,
    query(Q),
    Q,
    write_term(Q,[numbervars(true),quoted(true),double_quotes(true)]),
    write('.\n'),
    fail;
    halt.

File: let/input/beetle12.py
"""
beetle12.py – Encoding the Beetle‑12 example in **ProbLog**
=========================================================

ProbLog supports *annotated disjunctions*, which map neatly onto the
N3 “($ {...} {...} $)” choice constructs.  By giving each branch a
probability of 0.5 we capture the idea that *exactly one* of the options
holds in any particular possible world, while being able to compute the
**skeptical probability** (the probability that a query is true across
all possible choices).  In this deterministic example that probability
will turn out to be 1.

Install ProbLog with:

```bash
pip install problog
```

Run the script:

```bash
python beetle12.py
```

You should see that `beetle` is **beautiful** with probability **1.0**
and none of the intermediate colours/qualities are certain.
"""
from problog.program import PrologString
from problog import get_evaluatable

problog_code = """
% -------- facts -----------------------------------------------------------
car(beetle).

% -------- annotated disjunctions -----------------------------------------
1/2::prop(green,A); 1/2::prop(blue,A)  :- car(A).

1/2::prop(nice,A);  1/2::prop(pretty,A)  :- prop(green,A).

1/2::prop(pretty1,A); 1/2::prop(pretty2,A) :- prop(pretty,A).
1/2::prop(nice1,A);   1/2::prop(nice2,A)   :- prop(nice,A).

1/2::prop(pretty11,A); 1/2::prop(pretty12,A) :- prop(pretty1,A).
1/2::prop(pretty21,A); 1/2::prop(pretty22,A) :- prop(pretty2,A).

1/2::prop(nice11,A); 1/2::prop(nice12,A) :- prop(nice1,A).
1/2::prop(nice21,A); 1/2::prop(nice22,A) :- prop(nice2,A).

% -------- deterministic beauty rules -------------------------------------
prop(beautiful,A) :- prop(blue,A).
prop(beautiful,A) :- prop(pretty11,A).
prop(beautiful,A) :- prop(pretty12,A).
prop(beautiful,A) :- prop(pretty21,A).
prop(beautiful,A) :- prop(pretty22,A).
prop(beautiful,A) :- prop(nice11,A).
prop(beautiful,A) :- prop(nice12,A).
prop(beautiful,A) :- prop(nice21,A).
prop(beautiful,A) :- prop(nice22,A).

% -------- queries ---------------------------------------------------------
query(prop(beautiful,beetle)).
query(prop(green,beetle)).
query(prop(blue,beetle)).
"""

model = PrologString(problog_code)
result = get_evaluatable().create_from(model).evaluate()

for query, prob in result.items():
    print(f"{query}: {prob}")
